### PR TITLE
minisat: Move preregistration handling of new sat variables to theory level.

### PR DESCRIPTION
**Pull request by aniemetz**
The currently implemented behavior mirrors what was implemented in Minisat before.
Tested on whole SMT-LIB. Performance comparison is currently running.

### CODE REVIEW

File: src/prop/theory_preregistrar.cpp [hunk 1]
```diff
@@ -39,13 +67,46 @@ void TheoryPreregistrar::addAssertion(TNode n, TNode skolem, bool isLemma) {}
...
+    {
+      // cache for registration
+      d_sat_literals.emplace_back(n, d_env.getContext()->getLevel());
+    }
+  }
+}
+
+void TheoryPreregistrar::notifyBacktrack(uint32_t nlevels)
+{
+  (void)nlevels;
```
Comment (ajreynol) [inline src/prop/theory_preregistrar.cpp:87]: What is the purpose of this line?

**Review by ajreynol — COMMENTED** [review]
One minor question.

File: src/prop/theory_preregistrar.cpp (hunk 1 above)
Comment (aniemetz) [reply src/prop/theory_preregistrar.cpp:87]: To silence compiler warnings of unused variables.

Comment (aniemetz) [reply src/prop/theory_preregistrar.cpp:87]: Note that this argument was introduces in compliance with the new SAT CDCL(T) interface.

Comment (ajreynol) [reply src/prop/theory_preregistrar.cpp:87]: Yep, makes sense.

**Comment by ajreynol** [comment]
Pending performance tests, this LGTM.

**Comment by aniemetz** [comment]
Performance comparison, `2023-02-15-(non-)inc` is main, `2023-02-17-(non-)inc` is this PR.
```
                        status   total  solved     sat   unsat    best  timeout  memout  error  uniq  disagr   time_cpu     memory
config                                                                                                                            
2023-02-17-non-inc-cvc5     ok  391363  339823  151084  188739  254423    39720     234      0  2958       0 13325752.3 26747436.6
2023-02-15-non-inc-cvc5     ok  391363  337093  148793  188300  316340    42452     251      0   228       0 14096444.7 26826611.3
```
```
                    status  total   solved      sat    unsat   best  timeout  memout  error  uniq  disagr  time_cpu    memory
config                                                                                                                       
2023-02-17-inc-cvc5     ok  43285  5242250  4101944  1140306  29681     6464      76      0     2       0 2371722.6 7254143.5
2023-02-15-inc-cvc5     ok  43285  5099798  3984753  1115045  31371     6416      75      0     3       0 2346802.1 7313948.9
```

**Comment by aniemetz** [comment]
Per logic:
[CMakeCache.txt](https://github.com/cvc5/cvc5/files/10796254/CMakeCache.txt)
[cmp_inc_logics.txt](https://github.com/cvc5/cvc5/files/10796255/cmp_inc_logics.txt)
[cmp_non_inc_logics.txt](https://github.com/cvc5/cvc5/files/10796256/cmp_non_inc_logics.txt)

**Comment by aniemetz** [comment]
300s time limit, 8GB memory limit:
```
                             status   total  solved     sat   unsat    best  timeout  memout  error  uniq  disagr   time_cpu     memory
config                                                                                                                                 
2023-02-17-non-inc-cvc5          ok  391363  339823  151084  188739  307763    39720     234      0    35       0 13325752.3 26747436.6
2023-02-21-non-inc-cvc5-main     ok  391363  339811  151074  188737  307824    39737     233      0    23       0 13329983.5 26697377.0
```
```
                         status  total   solved      sat    unsat   best  timeout  memout  error  uniq  disagr  time_cpu    memory
config                                                                                                                            
2023-02-21-inc-cvc5-main     ok  43285  5332614  4173602  1159012  30790     6470      76      0     2       0 2372395.4 7258327.7
2023-02-17-inc-cvc5          ok  43285  5242250  4101944  1140306  31597     6464      76      0     3       0 2371722.6 7254143.5
```
Per logic:
[cmpr-inc-logics.txt](https://github.com/cvc5/cvc5/files/10807234/cmpr-inc-logics.txt)
[cmpr-non-inc-logics.txt](https://github.com/cvc5/cvc5/files/10807235/cmpr-non-inc-logics.txt)
On common instances:
[cmpr-non-inc-common.txt](https://github.com/cvc5/cvc5/files/10807241/cmpr-non-inc-common.txt)
Diff on non-inc is cluster noise, except for a handful (and of those, all but one are better than main: ~100+s vs time out):
[cmpr-non-inc-diff.txt](https://github.com/cvc5/cvc5/files/10807263/cmpr-non-inc-diff.txt)
For inc, the diff is limited to QF_ABVFP (cluster noise), QF_AUFLIA, QF_NIA, QF_UFLIA and QF_UFNIA. For the integer logics, I'm currently rerunning to rule out cluster noise.

**Review by ajreynol — APPROVED** [review]
LGTM, this is great.

**Comment by aniemetz** [comment]
Inc is all noise.

**Pull request merged**